source-braintree-native/source_braintree_native/api.py

```diff
@@ -0,0 +1,253 @@
+        """
+        Recursively convert a Braintree object and its nested objects to a dictionary.
+        Convenience objects intended to make subsequent Braintree requests easier are ommitted.
+        """
+        if isinstance(braintree_object, (Configuration, AddOnGateway, BraintreeGateway)):
+            return None
+        data = braintree_object.__dict__.copy()
+        # Remove convenience objects (like BraintreeGateways for making more requests).
+        for key in CONVENIENCE_OBJECTS:
+            data.pop(key, None)
+
+        for key, value in data.items():
+            if isinstance(value, AttributeGetter):
+                data[key] = _braintree_object_to_dict(value)
+            elif isinstance(value, datetime):
+                data[key] =  value.replace(tzinfo=UTC)
+            elif hasattr(value, "__dict__"):
+                data[key] = _braintree_object_to_dict(value)
+            elif isinstance(value, list):
+                new_value = []
+                for item in value:
+                    if hasattr(item, "__dict__"):
+                        new_value.append(_braintree_object_to_dict(item))
+                    elif isinstance(item, datetime):
+                        new_value.append(item.replace(tzinfo=UTC))
+                    else:
+                        new_value.append(item)
+
+                data[key] = new_value
+
+        # Remove private attributes.
+        data.pop('_setattrs', None)
+        return data
+
+
+async def snapshot_resources(
+        braintree_gateway: BraintreeGateway,
+        gateway_property: str,
+        gateway_response_field: str | None,
+        log: Logger,
+) -> AsyncGenerator[FullRefreshResource, None]:
+    resources = getattr(braintree_gateway, gateway_property).all()
+
+    iterator = getattr(resources, gateway_response_field) if gateway_response_field else resources
+    for object in iterator:
+        yield FullRefreshResource.model_validate(_braintree_object_to_dict(object))
+
+
+async def fetch_transactions(
+        braintree_gateway: BraintreeGateway,
+        log: Logger,
+        log_cursor: LogCursor,
+) -> AsyncGenerator[Transaction | LogCursor, None]:
+    assert isinstance(log_cursor, datetime)
+    most_recent_created_at = log_cursor
+    window_end = log_cursor + timedelta(days=WINDOW_SIZE)
+    end = min(window_end, datetime.now(tz=UTC))
+
+    collection = braintree_gateway.transaction.search(
+        TransactionSearch.created_at.between(log_cursor, end)
+    )
+
+    count = 0
+
+    for object in collection.items:
+        count += 1
+        doc = Transaction.model_validate(_braintree_object_to_dict(object))
+
+        if doc.created_at > log_cursor:
+            yield doc
+            most_recent_created_at = doc.created_at
+
+    if count >= TRANSACTION_SEARCH_LIMIT:
+        log.warning(f"{count} transactions returned in a single search which is greater than or equal to Braintree's documented maximum for a single transaction search.")
```

Comment:
Sanity checking my own understanding: Is the idea that this is non-fatal because there will be another regular backfill at some point in the future which will get any missing data?

Is there some subtle reason that a fresh backfill of the same resource won't hit this error?

Reply:
No you're right. A backfill will hit the same error. I originally included this log to trigger us to reduce the window size, but it makes more sense to make it a fatal error & let the user configure a different window size when it happens. I'll update this.